File: src/main/java/com/netsensia/rivalchess/model/Board.java
package com.netsensia.rivalchess.model;

import com.netsensia.rivalchess.model.util.BoardUtils;
import com.netsensia.rivalchess.model.util.FenUtils;
import com.netsensia.rivalchess.model.util.MoveMaker;

import java.util.ArrayList;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

public class Board {

    private final Map<Square, SquareOccupant> squareOccupants;
    private final Map<SquareOccupant, List<Square>> occupantSquares;

    private final int enPassantFile;

    private final boolean isWhiteKingSideCastleAvailable;
    private final boolean isWhiteQueenSideCastleAvailable;
    private final boolean isBlackKingSideCastleAvailable;
    private final boolean isBlackQueenSideCastleAvailable;

    private final int halfMoveCount;
    private final Colour sideToMove;

    private final int fullMoveCount;

    public Board(final BoardBuilder builder) {
        this.squareOccupants = new EnumMap<>(builder.squareOccupants);
        this.enPassantFile = builder.enPassantFile;
        this.isWhiteQueenSideCastleAvailable = builder.isWhiteQueenSideCastleAvailable;
        this.isBlackQueenSideCastleAvailable = builder.isBlackQueenSideCastleAvailable;
        this.isWhiteKingSideCastleAvailable = builder.isWhiteKingSideCastleAvailable;
        this.isBlackKingSideCastleAvailable = builder.isBlackKingSideCastleAvailable;
        this.halfMoveCount = builder.halfMoveCount;
        this.sideToMove = builder.sideToMove;
        this.fullMoveCount = builder.fullMoveCount;

        this.occupantSquares = new EnumMap<>(SquareOccupant.class);

        populateOccupantSquares();
    }

    public static Board fromFen(final String fen) {
        return FenUtils.getBoardModel(fen);
    }

    public Map<Square, SquareOccupant> getSquareOccupants() {
        return new EnumMap<>(squareOccupants);
    }

    public Board(final Board board) {

        squareOccupants = new EnumMap<>(board.getSquareOccupants());

        this.sideToMove = board.getSideToMove();
        this.enPassantFile = board.getEnPassantFile();

        this.isBlackKingSideCastleAvailable = board.isKingSideCastleAvailable(Colour.BLACK);
        this.isBlackQueenSideCastleAvailable = board.isQueenSideCastleAvailable(Colour.BLACK);
        this.isWhiteKingSideCastleAvailable = board.isKingSideCastleAvailable(Colour.WHITE);
        this.isWhiteQueenSideCastleAvailable = board.isQueenSideCastleAvailable(Colour.WHITE);

        this.halfMoveCount = board.getHalfMoveCount();
        this.fullMoveCount = board.getFullMoveCount();

        this.occupantSquares = new EnumMap<>(SquareOccupant.class);

        populateOccupantSquares();
    }

    private void populateOccupantSquares() {

        for (SquareOccupant so : SquareOccupant.values()) {
            occupantSquares.put(so, new ArrayList<>());
        }

        for (Square s : Square.values()) {
            final SquareOccupant so = squareOccupants.get(s);
            if (so != SquareOccupant.NONE) {
                occupantSquares.get(so).add(s);
            }
        }
    }

    public static Board fromMove(final Board board, final Move move) {
        return MoveMaker.makeMove(board, move);
    }

    public SquareOccupant getSquareOccupant(Square square) {
        return squareOccupants.get(square);
    }

    public boolean isKingSideCastleAvailable(final Colour colour) {
        return colour == Colour.WHITE ? isWhiteKingSideCastleAvailable : isBlackKingSideCastleAvailable;
    }

    public boolean isQueenSideCastleAvailable(final Colour colour) {
        return colour == Colour.WHITE ? isWhiteQueenSideCastleAvailable : isBlackQueenSideCastleAvailable;
    }

    public Stream<Map.Entry<Square, SquareOccupant>> squareOccupantStream() {
        return squareOccupants.entrySet().stream();
    }

    public List<Square> getSquaresWithOccupant(SquareOccupant squareOccupant) {
        return new ArrayList<>(occupantSquares.get(squareOccupant));
    }

    public int getHalfMoveCount() {
        return this.halfMoveCount;
    }

    public int getFullMoveCount() {
        return this.fullMoveCount;
    }

    public Colour getSideToMove() {
        return sideToMove;
    }

    public int getEnPassantFile() {
        return this.enPassantFile;
    }

    public List<Move> getLegalMoves() {
        return BoardUtils.getLegalMoves(this);
    }

    public boolean isCheck() {
        return BoardUtils.isCheck(this);
    }

    public static BoardBuilder builder() {
        return new BoardBuilder();
    }

    public boolean isSquareAttackedBy(final Square square, final Colour byColour) {
        return BoardUtils.isSquareAttackedBy(this, square, byColour);
    }

    @Override
    public boolean equals(Object o) {

        if (o instanceof Board) {
            Board bo = (Board) o;

            for (Square square : Square.values()) {
                if (getSquareOccupant(square) != bo.getSquareOccupant(square)) {
                    return false;
                }
            }

            return  this.getEnPassantFile() == bo.getEnPassantFile() &&
                    this.getSideToMove() == bo.getSideToMove() &&
                    this.getHalfMoveCount() == bo.getHalfMoveCount() &&
                    this.getFullMoveCount() == bo.getFullMoveCount() &&
                    this.isKingSideCastleAvailable(Colour.WHITE) == bo.isKingSideCastleAvailable(Colour.WHITE) &&
                    this.isKingSideCastleAvailable(Colour.BLACK) == bo.isKingSideCastleAvailable(Colour.BLACK) &&
                    this.isQueenSideCastleAvailable(Colour.WHITE) == bo.isQueenSideCastleAvailable(Colour.WHITE) &&
                    this.isQueenSideCastleAvailable(Colour.BLACK) == bo.isQueenSideCastleAvailable(Colour.BLACK);
        }

        return false;
    }

    @Override
    public int hashCode() {
        return this.toString().hashCode();
    }

    @Override
    public String toString() {
        StringBuilder s = new StringBuilder();

        s.append(squareOccupants);

        s.append("Castle privileges: " +
                isWhiteKingSideCastleAvailable +
                isWhiteQueenSideCastleAvailable +
                isBlackKingSideCastleAvailable +
                isBlackQueenSideCastleAvailable);
        s.append("\n");
        s.append("En passant file: " + enPassantFile);
        s.append("\n");
        s.append("Half move count: " + halfMoveCount);

        return s.toString();
    }

    public static class BoardBuilder
    {
        private EnumMap<Square, SquareOccupant> squareOccupants;

        private int enPassantFile;

        private boolean isWhiteKingSideCastleAvailable;
        private boolean isWhiteQueenSideCastleAvailable;
        private boolean isBlackKingSideCastleAvailable;
        private boolean isBlackQueenSideCastleAvailable;

        private int halfMoveCount;
        private int fullMoveCount;

        private Colour sideToMove;

        private BoardBuilder() {
            enPassantFile = -1;
            isBlackQueenSideCastleAvailable = false;
            isWhiteQueenSideCastleAvailable = false;
            isWhiteKingSideCastleAvailable = false;
            isBlackKingSideCastleAvailable = false;
            halfMoveCount = 0;
            sideToMove = Colour.WHITE;

            squareOccupants = new EnumMap<>(Square.class);
            for (Square square : Square.values()) {
                squareOccupants.put(
                        square,
                        SquareOccupant.NONE
                );
            }
        }

        public BoardBuilder(final Board board) {
            squareOccupants = new EnumMap<>(board.getSquareOccupants());
            enPassantFile = board.getEnPassantFile();
            isWhiteKingSideCastleAvailable = board.isKingSideCastleAvailable(Colour.WHITE);
            isBlackKingSideCastleAvailable = board.isKingSideCastleAvailable(Colour.BLACK);
            isWhiteQueenSideCastleAvailable = board.isQueenSideCastleAvailable(Colour.WHITE);
            isBlackQueenSideCastleAvailable = board.isQueenSideCastleAvailable(Colour.BLACK);
            halfMoveCount = board.getHalfMoveCount();
            sideToMove = board.getSideToMove();
            fullMoveCount = board.getFullMoveCount();
        }

        public BoardBuilder withSquareOccupant(final Square square, final SquareOccupant squareOccupant) {
            this.squareOccupants.put(square, squareOccupant);
            return this;
        }

        public BoardBuilder withEnPassantFile(final int enPassantFile) {
            this.enPassantFile = enPassantFile;
            return this;
        }

        public BoardBuilder withFullMoveCount(final int fullMoveCount) {
            this.fullMoveCount = fullMoveCount;
            return this;
        }

        public BoardBuilder withHalfMoveCount(final int halfMoveCount) {
            this.halfMoveCount = halfMoveCount;
            return this;
        }

        public BoardBuilder withSideToMove(final Colour sideToMove) {
            this.sideToMove = sideToMove;
            return this;
        }

        public BoardBuilder withIsWhiteKingSideCastleAvailable(final boolean isWhiteKingSideCastleAvailable) {
            this.isWhiteKingSideCastleAvailable = isWhiteKingSideCastleAvailable;
            return this;
        }

        public BoardBuilder withIsBlackKingSideCastleAvailable(final boolean isBlackKingSideCastleAvailable) {
            this.isBlackKingSideCastleAvailable = isBlackKingSideCastleAvailable;
            return this;
        }

        public BoardBuilder withIsWhiteQueenSideCastleAvailable(final boolean isWhiteQueenSideCastleAvailable) {
            this.isWhiteQueenSideCastleAvailable = isWhiteQueenSideCastleAvailable;
            return this;
        }

        public BoardBuilder withIsBlackQueenSideCastleAvailable(final boolean isBlackQueenSideCastleAvailable) {
            this.isBlackQueenSideCastleAvailable = isBlackQueenSideCastleAvailable;
            return this;
        }

        public BoardBuilder withIsQueenSideCastleAvailable(final Colour colour, final boolean isQueenSideCastleAvailable) {
            if (colour == Colour.WHITE) {
                this.isWhiteQueenSideCastleAvailable = isQueenSideCastleAvailable;
            } else {
                this.isBlackQueenSideCastleAvailable = isQueenSideCastleAvailable;
            }
            return this;
        }

        public BoardBuilder withIsKingSideCastleAvailable(final Colour colour, final boolean isKingSideCastleAvailable) {
            if (colour == Colour.WHITE) {
                this.isWhiteKingSideCastleAvailable = isKingSideCastleAvailable;
            } else {
                this.isBlackKingSideCastleAvailable = isKingSideCastleAvailable;
            }
            return this;
        }

        public BoardBuilder withNoCastleFlags() {
            this.isWhiteQueenSideCastleAvailable = false;
            this.isBlackQueenSideCastleAvailable = false;
            this.isWhiteKingSideCastleAvailable = false;
            this.isBlackKingSideCastleAvailable = false;
            return this;
        }

        public Board build() {
            return new Board(this);
        }

    }
}

File: src/main/java/com/netsensia/rivalchess/model/Board.kt
package com.netsensia.rivalchess.model

import com.netsensia.rivalchess.model.util.BoardUtils
import com.netsensia.rivalchess.model.util.FenUtils
import com.netsensia.rivalchess.model.util.MoveMaker
import java.util.*

class Board {
    private val squareOccupants: Map<Square, SquareOccupant>
    private val occupantSquares: MutableMap<SquareOccupant, MutableList<Square>>
    val enPassantFile: Int
    private val isWhiteKingSideCastleAvailable: Boolean
    private val isWhiteQueenSideCastleAvailable: Boolean
    private val isBlackKingSideCastleAvailable: Boolean
    private val isBlackQueenSideCastleAvailable: Boolean
    val halfMoveCount: Int
    val sideToMove: Colour
    val fullMoveCount: Int

    constructor(builder: BoardBuilder) {
        squareOccupants = EnumMap(builder.squareOccupants)
        enPassantFile = builder.enPassantFile
        isWhiteQueenSideCastleAvailable = builder.isWhiteQueenSideCastleAvailable
        isBlackQueenSideCastleAvailable = builder.isBlackQueenSideCastleAvailable
        isWhiteKingSideCastleAvailable = builder.isWhiteKingSideCastleAvailable
        isBlackKingSideCastleAvailable = builder.isBlackKingSideCastleAvailable
        halfMoveCount = builder.halfMoveCount
        sideToMove = builder.sideToMove
        fullMoveCount = builder.fullMoveCount
        occupantSquares = EnumMap(SquareOccupant::class.java)
        populateOccupantSquares()
    }

    fun getSquareOccupants(): Map<Square, SquareOccupant> {
        return EnumMap(squareOccupants)
    }

    constructor(board: Board) {
        squareOccupants = EnumMap(board.getSquareOccupants())
        sideToMove = board.sideToMove
        enPassantFile = board.enPassantFile
        isBlackKingSideCastleAvailable = board.isKingSideCastleAvailable(Colour.BLACK)
        isBlackQueenSideCastleAvailable = board.isQueenSideCastleAvailable(Colour.BLACK)
        isWhiteKingSideCastleAvailable = board.isKingSideCastleAvailable(Colour.WHITE)
        isWhiteQueenSideCastleAvailable = board.isQueenSideCastleAvailable(Colour.WHITE)
        halfMoveCount = board.halfMoveCount
        fullMoveCount = board.fullMoveCount
        occupantSquares = EnumMap(SquareOccupant::class.java)
        populateOccupantSquares()
    }

    private fun populateOccupantSquares() {
        for (so in SquareOccupant.values()) {
            occupantSquares[so] = ArrayList()
        }
        for (s in Square.values()) {
            val so = squareOccupants[s]
            if (so != SquareOccupant.NONE) {
                occupantSquares[so]!!.add(s)
            }
        }
    }

    fun getSquareOccupant(square: Square): SquareOccupant {
        return squareOccupants[square]!!
    }

    fun isKingSideCastleAvailable(colour: Colour): Boolean {
        return if (colour == Colour.WHITE) isWhiteKingSideCastleAvailable else isBlackKingSideCastleAvailable
    }

    fun isQueenSideCastleAvailable(colour: Colour): Boolean {
        return if (colour == Colour.WHITE) isWhiteQueenSideCastleAvailable else isBlackQueenSideCastleAvailable
    }

    fun getSquaresWithOccupant(squareOccupant: SquareOccupant): List<Square> {
        return ArrayList(occupantSquares[squareOccupant])
    }

    val legalMoves: List<Move>
        get() = BoardUtils.getLegalMoves(this)

    override fun equals(other: Any?): Boolean {
        if (other is Board) {
            val bo = other
            for (square in Square.values()) {
                if (getSquareOccupant(square) != bo.getSquareOccupant(square)) {
                    return false
                }
            }
            return enPassantFile == bo.enPassantFile &&
                    sideToMove == bo.sideToMove &&
                    halfMoveCount == bo.halfMoveCount &&
                    fullMoveCount == bo.fullMoveCount &&
                    isKingSideCastleAvailable(Colour.WHITE) == bo.isKingSideCastleAvailable(Colour.WHITE) &&
                    isKingSideCastleAvailable(Colour.BLACK) == bo.isKingSideCastleAvailable(Colour.BLACK) &&
                    isQueenSideCastleAvailable(Colour.WHITE) == bo.isQueenSideCastleAvailable(Colour.WHITE) &&
                    isQueenSideCastleAvailable(Colour.BLACK) == bo.isQueenSideCastleAvailable(Colour.BLACK)
        }
        return false
    }

    override fun hashCode(): Int {
        return this.toString().hashCode()
    }

    override fun toString(): String {
        val s = StringBuilder()
        s.append(squareOccupants)
        s.append("Castle privileges: " +
                isWhiteKingSideCastleAvailable +
                isWhiteQueenSideCastleAvailable +
                isBlackKingSideCastleAvailable +
                isBlackQueenSideCastleAvailable)
        s.append("\n")
        s.append("En passant file: $enPassantFile")
        s.append("\n")
        s.append("Half move count: $halfMoveCount")
        return s.toString()
    }

    class BoardBuilder {
        var squareOccupants: EnumMap<Square, SquareOccupant>
        var enPassantFile: Int
        var isWhiteKingSideCastleAvailable: Boolean
        var isWhiteQueenSideCastleAvailable: Boolean
        var isBlackKingSideCastleAvailable: Boolean
        var isBlackQueenSideCastleAvailable: Boolean
        var halfMoveCount: Int
        var fullMoveCount = 0
        var sideToMove: Colour

        constructor() {
            enPassantFile = -1
            isBlackQueenSideCastleAvailable = false
            isWhiteQueenSideCastleAvailable = false
            isWhiteKingSideCastleAvailable = false
            isBlackKingSideCastleAvailable = false
            halfMoveCount = 0
            sideToMove = Colour.WHITE
            squareOccupants = EnumMap(Square::class.java)
            for (square in Square.values()) {
                squareOccupants[square] = SquareOccupant.NONE
            }
        }

        constructor(board: Board) {
            squareOccupants = EnumMap(board.getSquareOccupants())
            enPassantFile = board.enPassantFile
            isWhiteKingSideCastleAvailable = board.isKingSideCastleAvailable(Colour.WHITE)
            isBlackKingSideCastleAvailable = board.isKingSideCastleAvailable(Colour.BLACK)
            isWhiteQueenSideCastleAvailable = board.isQueenSideCastleAvailable(Colour.WHITE)
            isBlackQueenSideCastleAvailable = board.isQueenSideCastleAvailable(Colour.BLACK)
            halfMoveCount = board.halfMoveCount
            sideToMove = board.sideToMove
            fullMoveCount = board.fullMoveCount
        }

        fun withSquareOccupant(square: Square, squareOccupant: SquareOccupant) = apply {
            squareOccupants[square] = squareOccupant
        }

        fun withEnPassantFile(enPassantFile: Int) = apply {
            this.enPassantFile = enPassantFile
        }

        fun withFullMoveCount(fullMoveCount: Int) = apply {
            this.fullMoveCount = fullMoveCount
        }

        fun withHalfMoveCount(halfMoveCount: Int) = apply {
            this.halfMoveCount = halfMoveCount
            return this
        }

        fun withSideToMove(sideToMove: Colour) = apply {
            this.sideToMove = sideToMove
        }

        fun withIsWhiteKingSideCastleAvailable(isWhiteKingSideCastleAvailable: Boolean) = apply {
            this.isWhiteKingSideCastleAvailable = isWhiteKingSideCastleAvailable
        }

        fun withIsBlackKingSideCastleAvailable(isBlackKingSideCastleAvailable: Boolean) = apply {
            this.isBlackKingSideCastleAvailable = isBlackKingSideCastleAvailable
        }

        fun withIsWhiteQueenSideCastleAvailable(isWhiteQueenSideCastleAvailable: Boolean) = apply {
            this.isWhiteQueenSideCastleAvailable = isWhiteQueenSideCastleAvailable
        }

        fun withIsBlackQueenSideCastleAvailable(isBlackQueenSideCastleAvailable: Boolean) = apply {
            this.isBlackQueenSideCastleAvailable = isBlackQueenSideCastleAvailable
        }

        fun withIsQueenSideCastleAvailable(colour: Colour, isQueenSideCastleAvailable: Boolean) = apply {
            if (colour == Colour.WHITE) {
                isWhiteQueenSideCastleAvailable = isQueenSideCastleAvailable
            } else {
                isBlackQueenSideCastleAvailable = isQueenSideCastleAvailable
            }
            return this
        }

        fun withIsKingSideCastleAvailable(colour: Colour, isKingSideCastleAvailable: Boolean) = apply {
            if (colour == Colour.WHITE) {
                isWhiteKingSideCastleAvailable = isKingSideCastleAvailable
            } else {
                isBlackKingSideCastleAvailable = isKingSideCastleAvailable
            }
        }

        fun withNoCastleFlags() = apply {
            isWhiteQueenSideCastleAvailable = false
            isBlackQueenSideCastleAvailable = false
            isWhiteKingSideCastleAvailable = false
            isBlackKingSideCastleAvailable = false
        }

        fun build(): Board {
            return Board(this)
        }
    }

    companion object {
        @kotlin.jvm.JvmStatic
        fun fromFen(fen: String): Board {
            return FenUtils.getBoardModel(fen)
        }

        @kotlin.jvm.JvmStatic
        fun fromMove(board: Board, move: Move): Board {
            return MoveMaker.makeMove(board, move)
        }

        @kotlin.jvm.JvmStatic
        fun builder(): BoardBuilder {
            return BoardBuilder()
        }
    }
}